// @arco-design/web-vue/lib/tree/hooks/use-tree-data.js
"use strict";
var vue = require("vue");
var index = require("../utils/index.js");
var treeData = require("../utils/tree-data.js");
function useTreeData(props) {
  const {
    treeData: propTreeData,
    fieldNames,
    selectable,
    showLine,
    blockNode,
    checkable,
    loadMore,
    draggable
  } = vue.toRefs(props);
  const treeData$1 = vue.ref([]);
  vue.watchEffect(() => {
    treeData$1.value = treeData.generateTreeData(propTreeData.value || [], {
      selectable: !!(selectable == null ? void 0 : selectable.value),
      showLine: !!(showLine == null ? void 0 : showLine.value),
      blockNode: !!(blockNode == null ? void 0 : blockNode.value),
      checkable: !!(checkable == null ? void 0 : checkable.value),
      fieldNames: fieldNames == null ? void 0 : fieldNames.value,
      loadMore: !!(loadMore == null ? void 0 : loadMore.value),
      draggable: !!(draggable == null ? void 0 : draggable.value)
    });
  });
  const flattenTreeData = vue.computed(() => index.getFlattenTreeData(treeData$1.value));
  const key2TreeNode = vue.computed(() => index.getKey2TreeNode(flattenTreeData.value));
  return { treeData: treeData$1, flattenTreeData, key2TreeNode };
}
module.exports = useTreeData;
