// @arco-design/web-vue/lib/checkbox/checkbox-group.js
"use strict";
var vue = require("vue");
var globalConfig = require("../_utils/global-config.js");
var constant = require("../_utils/constant.js");
var context = require("./context.js");
var pluginVue_exportHelper = require("../_virtual/plugin-vue_export-helper");
const _sfc_main = vue.defineComponent({
  name: "CheckboxGroup",
  props: {
    modelValue: {
      type: Array,
      default: void 0
    },
    defaultValue: {
      type: Array,
      default: () => []
    },
    direction: {
      type: String,
      default: "horizontal",
      validator: (value) => constant.DIRECTIONS.includes(value)
    },
    disabled: {
      type: Boolean,
      default: false
    },
    onChange: {
      type: [Function, Array]
    }
  },
  emits: [
    "update:modelValue",
    "change"
  ],
  setup(props, { emit }) {
    const { disabled } = vue.toRefs(props);
    const prefixCls = globalConfig.getPrefixCls("checkbox-group");
    const _value = vue.ref(props.defaultValue);
    const computedValue = vue.computed(() => {
      var _a;
      return (_a = props.modelValue) != null ? _a : _value.value;
    });
    const handleChange = (value, e) => {
      _value.value = value;
      emit("update:modelValue", value);
      vue.nextTick(() => {
        emit("change", value, e);
      });
    };
    vue.provide(context.checkboxGroupKey, vue.reactive({
      name: "ArcoCheckboxGroup",
      computedValue,
      disabled,
      handleChange
    }));
    const cls = vue.computed(() => [
      prefixCls,
      `${prefixCls}-direction-${props.direction}`
    ]);
    vue.watch(() => props.modelValue, (curValue) => {
      if (curValue) {
        _value.value = [...curValue];
      }
    });
    return {
      prefixCls,
      cls
    };
  }
});
function _sfc_render(_ctx, _cache, $props, $setup, $data, $options) {
  return vue.openBlock(), vue.createElementBlock("span", {
    class: vue.normalizeClass(_ctx.cls)
  }, [
    vue.renderSlot(_ctx.$slots, "default")
  ], 2);
}
var CheckboxGroup = /* @__PURE__ */ pluginVue_exportHelper(_sfc_main, [["render", _sfc_render]]);
module.exports = CheckboxGroup;
