// @arco-design/web-vue/es/cascader/cascader-panel.js
import { defineComponent, createVNode, TransitionGroup, isVNode } from "vue";
import CascaderOption from "./cascader-option.js";
import { getPrefixCls } from "../_utils/global-config.js";
import Empty from "../empty/index.js";
import Spin from "../spin/index.js";
function _isSlot(s) {
  return typeof s === "function" || Object.prototype.toString.call(s) === "[object Object]" && !isVNode(s);
}
var CascaderPanel = defineComponent({
  name: "CascaderPanel",
  props: {
    displayColumns: {
      type: Array,
      required: true
    },
    selectedPath: {
      type: Array,
      required: true
    },
    activeKey: String,
    computedKeys: {
      type: Array,
      required: true
    },
    totalLevel: {
      type: Number,
      required: true
    },
    expandTrigger: String,
    multiple: Boolean,
    checkStrictly: Boolean,
    loading: Boolean
  },
  setup(props) {
    const prefixCls = getPrefixCls("cascader");
    const renderEmpty = () => {
      return createVNode(Empty, null, null);
    };
    const renderColumn = (column, level = 0) => {
      return createVNode("div", {
        "class": `${prefixCls}-panel-column`,
        "key": `column-${level}`,
        "style": {
          zIndex: props.totalLevel - level
        }
      }, [column.length === 0 ? createVNode("div", {
        "class": `${prefixCls}-list-empty`
      }, [renderEmpty()]) : createVNode("ul", {
        "class": [`${prefixCls}-list`, {
          [`${prefixCls}-list-multiple`]: Boolean(props == null ? void 0 : props.multiple),
          [`${prefixCls}-list-strictly`]: Boolean(props == null ? void 0 : props.checkStrictly)
        }]
      }, [column.map((item) => {
        return createVNode(CascaderOption, {
          "key": item.key,
          "option": item,
          "computedKeys": props.computedKeys,
          "active": props.selectedPath.includes(item.key) || item.key === props.activeKey,
          "multiple": props.multiple,
          "checkStrictly": props.checkStrictly,
          "expandTrigger": props.expandTrigger
        }, null);
      })])]);
    };
    const renderContent = () => {
      if (props.loading) {
        return createVNode("div", {
          "class": [`${prefixCls}-panel-column`, `${prefixCls}-panel-column-loading`]
        }, [createVNode(Spin, null, null)]);
      }
      if (props.displayColumns.length === 0) {
        return createVNode("div", {
          "class": `${prefixCls}-panel-column`
        }, [createVNode("div", {
          "class": `${prefixCls}-list-empty`
        }, [renderEmpty()])]);
      }
      return props.displayColumns.map((column, index) => renderColumn(column, index));
    };
    return () => {
      let _slot;
      return createVNode(TransitionGroup, {
        "tag": "div",
        "name": "cascader-slide",
        "class": `${prefixCls}-panel`
      }, _isSlot(_slot = renderContent()) ? _slot : {
        default: () => [_slot]
      });
    };
  }
});
export { CascaderPanel as default };
