// @arco-design/web-vue/es/cascader/cascader-search-panel.js
import { defineComponent, createVNode } from "vue";
import { getPrefixCls } from "../_utils/global-config.js";
import Empty from "../empty/index.js";
import Spin from "../spin/index.js";
import CascaderOption from "./cascader-option.js";
var CascaderSearchPanel = defineComponent({
  name: "CascaderSearchPanel",
  props: {
    options: {
      type: Array,
      required: true
    },
    computedKeys: {
      type: Array,
      required: true
    },
    loading: Boolean,
    activeKey: String,
    multiple: Boolean
  },
  setup(props) {
    const prefixCls = getPrefixCls("cascader");
    const renderContent = () => {
      if (props.loading) {
        return createVNode(Spin, null, null);
      }
      if (props.options.length === 0) {
        return createVNode("div", {
          "class": `${prefixCls}-list-empty`
        }, [createVNode(Empty, null, null)]);
      }
      return createVNode("ul", {
        "class": [`${prefixCls}-list`, `${prefixCls}-search-list`, {
          [`${prefixCls}-list-multiple`]: props.multiple
        }]
      }, [props.options.map((item) => createVNode(CascaderOption, {
        "key": item.key,
        "class": `${prefixCls}-search-option`,
        "option": item,
        "computedKeys": props.computedKeys,
        "active": item.key === props.activeKey,
        "multiple": props.multiple,
        "searchOption": true
      }, null))]);
    };
    return () => createVNode("div", {
      "class": [`${prefixCls}-panel`, `${prefixCls}-search-panel`]
    }, [renderContent()]);
  }
});
export { CascaderSearchPanel as default };
