// @arco-design/web-vue/lib/cascader/cascader-search-panel.js
"use strict";
var vue = require("vue");
var globalConfig = require("../_utils/global-config.js");
var index$1 = require("../empty/index.js");
var index = require("../spin/index.js");
var cascaderOption = require("./cascader-option.js");
var CascaderSearchPanel = vue.defineComponent({
  name: "CascaderSearchPanel",
  props: {
    options: {
      type: Array,
      required: true
    },
    computedKeys: {
      type: Array,
      required: true
    },
    loading: Boolean,
    activeKey: String,
    multiple: Boolean
  },
  setup(props) {
    const prefixCls = globalConfig.getPrefixCls("cascader");
    const renderContent = () => {
      if (props.loading) {
        return vue.createVNode(index, null, null);
      }
      if (props.options.length === 0) {
        return vue.createVNode("div", {
          "class": `${prefixCls}-list-empty`
        }, [vue.createVNode(index$1, null, null)]);
      }
      return vue.createVNode("ul", {
        "class": [`${prefixCls}-list`, `${prefixCls}-search-list`, {
          [`${prefixCls}-list-multiple`]: props.multiple
        }]
      }, [props.options.map((item) => vue.createVNode(cascaderOption, {
        "key": item.key,
        "class": `${prefixCls}-search-option`,
        "option": item,
        "computedKeys": props.computedKeys,
        "active": item.key === props.activeKey,
        "multiple": props.multiple,
        "searchOption": true
      }, null))]);
    };
    return () => vue.createVNode("div", {
      "class": [`${prefixCls}-panel`, `${prefixCls}-search-panel`]
    }, [renderContent()]);
  }
});
module.exports = CascaderSearchPanel;
