// @arco-design/web-vue/lib/radio/radio-group.js
"use strict";
var vue = require("vue");
var globalConfig = require("../_utils/global-config.js");
var constant = require("../_utils/constant.js");
var context = require("./context.js");
var pluginVue_exportHelper = require("../_virtual/plugin-vue_export-helper");
const _sfc_main = vue.defineComponent({
  name: "RadioGroup",
  props: {
    modelValue: {
      type: [String, Number, Boolean],
      default: void 0
    },
    defaultValue: {
      type: [String, Number, Boolean],
      default: ""
    },
    type: {
      type: String,
      default: "radio",
      validator: (value) => context.RADIO_TYPES.includes(value)
    },
    size: {
      type: String,
      default: "medium",
      validator: (value) => constant.SIZES.includes(value)
    },
    direction: {
      type: String,
      default: "horizontal",
      validator: (value) => constant.DIRECTIONS.includes(value)
    },
    disabled: {
      type: Boolean,
      default: false
    },
    onChange: {
      type: [Function, Array]
    }
  },
  emits: [
    "update:modelValue",
    "change"
  ],
  setup(props, { emit }) {
    const prefixCls = globalConfig.getPrefixCls("radio-group");
    const { size, type, disabled, modelValue } = vue.toRefs(props);
    const _value = vue.ref(props.defaultValue);
    const computedValue = vue.computed(() => {
      var _a;
      return (_a = props.modelValue) != null ? _a : _value.value;
    });
    const handleChange = (value, e) => {
      _value.value = value;
      emit("update:modelValue", value);
      vue.nextTick(() => {
        emit("change", value, e);
      });
    };
    vue.provide(context.radioGroupKey, vue.reactive({
      name: "ArcoRadioGroup",
      value: computedValue,
      size,
      type,
      disabled,
      handleChange
    }));
    vue.watch(computedValue, (cur) => {
      if (_value.value !== cur) {
        _value.value = cur;
      }
    });
    const cls = vue.computed(() => [
      `${prefixCls}${props.type === "button" ? "-button" : ""}`,
      `${prefixCls}-size-${props.size}`,
      `${prefixCls}-direction-${props.direction}`,
      {
        [`${prefixCls}-disabled`]: props.disabled
      }
    ]);
    return {
      prefixCls,
      cls
    };
  }
});
function _sfc_render(_ctx, _cache, $props, $setup, $data, $options) {
  return vue.openBlock(), vue.createElementBlock("span", {
    class: vue.normalizeClass(_ctx.cls)
  }, [
    vue.renderSlot(_ctx.$slots, "default")
  ], 2);
}
var RadioGroup = /* @__PURE__ */ pluginVue_exportHelper(_sfc_main, [["render", _sfc_render]]);
module.exports = RadioGroup;
