// @arco-design/web-vue/lib/input/input-search.js
"use strict";
var vue = require("vue");
var globalConfig = require("../_utils/global-config.js");
var iconHover = require("../_components/icon-hover.js");
var index$1 = require("../icon/icon-search/index.js");
var index = require("../icon/icon-loading/index.js");
var index$2 = require("../button/index.js");
var input = require("./input.js");
var context = require("../config-provider/context.js");
var InputSearch = vue.defineComponent({
  name: "InputSearch",
  props: {
    searchButton: {
      type: Boolean,
      default: false
    },
    loading: {
      type: Boolean,
      default: false
    },
    disabled: {
      type: Boolean,
      default: false
    },
    size: {
      type: String,
      default: () => {
        var _a, _b;
        return (_b = (_a = vue.inject(context.configProviderInjectionKey, void 0)) == null ? void 0 : _a.size) != null ? _b : "medium";
      }
    },
    buttonProps: {
      type: Object
    },
    onSearch: {
      type: [Function, Array]
    }
  },
  emits: [
    "search"
  ],
  setup(props, {
    emit,
    slots
  }) {
    const prefixCls = globalConfig.getPrefixCls("input-search");
    const inputRef = vue.ref();
    const handleClick = (e) => {
      if (inputRef.value.inputRef) {
        emit("search", inputRef.value.inputRef.value, e);
      }
    };
    const renderSuffix = () => {
      var _a;
      return vue.createVNode(vue.Fragment, null, [props.loading ? vue.createVNode(index, null, null) : vue.createVNode(iconHover, {
        "onClick": handleClick
      }, {
        default: () => [vue.createVNode(index$1, null, null)]
      }), (_a = slots.suffix) == null ? void 0 : _a.call(slots)]);
    };
    const renderButton = () => {
      return vue.createVNode(index$2["default"], vue.mergeProps({
        "type": "primary",
        "class": `${prefixCls}-btn`,
        "disabled": props.disabled,
        "size": props.size,
        "loading": props.loading
      }, props.buttonProps, {
        "onClick": handleClick
      }), {
        icon: () => vue.createVNode(index$1, null, null)
      });
    };
    const render = () => vue.createVNode(input, {
      "ref": inputRef,
      "class": prefixCls,
      "size": props.size,
      "disabled": props.disabled
    }, {
      prepend: slots.prepend,
      prefix: slots.prefix,
      suffix: props.searchButton ? slots.suffix : renderSuffix,
      append: props.searchButton ? renderButton : slots.append
    });
    return {
      inputRef,
      render
    };
  },
  methods: {
    focus() {
      var _a;
      (_a = this.inputRef) == null ? void 0 : _a.focus();
    },
    blur() {
      var _a;
      (_a = this.inputRef) == null ? void 0 : _a.blur();
    }
  },
  render() {
    return this.render();
  }
});
module.exports = InputSearch;
