// @arco-design/web-vue/lib/input/input.js
"use strict";
var vue = require("vue");
var globalConfig = require("../_utils/global-config.js");
var constant = require("../_utils/constant.js");
var keycode = require("../_utils/keycode.js");
var iconHover = require("../_components/icon-hover.js");
var index = require("../icon/icon-close/index.js");
var omit = require("../_utils/omit.js");
var pick = require("../_utils/pick.js");
var is = require("../_utils/is.js");
var context = require("../config-provider/context.js");
var _Input = vue.defineComponent({
  name: "Input",
  inheritAttrs: false,
  props: {
    modelValue: String,
    defaultValue: {
      type: String,
      default: ""
    },
    size: {
      type: String,
      default: () => {
        var _a, _b;
        return (_b = (_a = vue.inject(context.configProviderInjectionKey, void 0)) == null ? void 0 : _a.size) != null ? _b : "medium";
      }
    },
    allowClear: {
      type: Boolean,
      default: false
    },
    disabled: {
      type: Boolean,
      default: false
    },
    readonly: {
      type: Boolean,
      default: false
    },
    error: {
      type: Boolean,
      default: false
    },
    placeholder: String,
    maxLength: {
      type: [Number, Object],
      default: 0
    },
    showWordLimit: {
      type: Boolean,
      default: false
    },
    wordLength: {
      type: Function
    },
    wordSlice: {
      type: Function
    },
    type: {
      type: String,
      default: "text"
    },
    onInput: {
      type: [Function, Array]
    },
    onChange: {
      type: [Function, Array]
    },
    onPressEnter: {
      type: [Function, Array]
    },
    onClear: {
      type: [Function, Array]
    },
    onFocus: {
      type: [Function, Array]
    },
    onBlur: {
      type: [Function, Array]
    }
  },
  emits: [
    "update:modelValue",
    "input",
    "change",
    "pressEnter",
    "clear",
    "focus",
    "blur"
  ],
  setup(props, {
    emit,
    slots,
    attrs
  }) {
    const prefixCls = globalConfig.getPrefixCls("input");
    const inputRef = vue.ref();
    const _value = vue.ref(props.defaultValue);
    const computedValue = vue.computed(() => {
      var _a;
      return (_a = props.modelValue) != null ? _a : _value.value;
    });
    let preValue = computedValue.value;
    const focused = vue.ref(false);
    const showClearBtn = vue.computed(() => props.allowClear && !props.disabled && Boolean(computedValue.value));
    const isComposition = vue.ref(false);
    const compositionValue = vue.ref("");
    const getValueLength = (value) => {
      var _a;
      if (is.isFunction(props.wordLength)) {
        return props.wordLength(value);
      }
      return (_a = value.length) != null ? _a : 0;
    };
    const valueLength = vue.computed(() => getValueLength(computedValue.value));
    const mergedError = vue.computed(() => {
      if (props.error) {
        return props.error;
      }
      return Boolean(is.isObject(props.maxLength) && props.maxLength.errorOnly && valueLength.value > maxLength.value);
    });
    const maxLengthErrorOnly = vue.computed(() => is.isObject(props.maxLength) && Boolean(props.maxLength.errorOnly));
    const maxLength = vue.computed(() => {
      if (is.isObject(props.maxLength)) {
        return props.maxLength.length;
      }
      return props.maxLength;
    });
    const updateValue = (value, inner = true) => {
      var _a, _b;
      if (maxLength.value && !maxLengthErrorOnly.value && getValueLength(value) > maxLength.value) {
        value = (_b = (_a = props.wordSlice) == null ? void 0 : _a.call(props, value, maxLength.value)) != null ? _b : value.slice(0, maxLength.value);
      }
      _value.value = value;
      if (inner) {
        emit("update:modelValue", value);
      }
    };
    const handleMousedown = (e) => {
      if (inputRef.value && e.target !== inputRef.value) {
        e.preventDefault();
        inputRef.value.focus();
      }
    };
    const emitChange = (value, ev) => {
      if (value !== preValue) {
        preValue = value;
        emit("change", value, ev);
      }
    };
    const handleFocus = (e) => {
      focused.value = true;
      preValue = computedValue.value;
      emit("focus", e);
    };
    const handleBlur = (e) => {
      focused.value = false;
      emitChange(computedValue.value, e);
      emit("blur", e);
    };
    const handleComposition = (e) => {
      var _a;
      const {
        value
      } = e.target;
      if (e.type === "compositionend") {
        isComposition.value = false;
        compositionValue.value = "";
        updateValue(value);
        emit("input", value, e);
        vue.nextTick(() => {
          if (inputRef.value && computedValue.value !== inputRef.value.value) {
            inputRef.value.value = computedValue.value;
          }
        });
      } else {
        isComposition.value = true;
        compositionValue.value = computedValue.value + ((_a = e.data) != null ? _a : "");
      }
    };
    const handleInput = (e) => {
      const {
        value
      } = e.target;
      if (!isComposition.value) {
        updateValue(value);
        emit("input", value, e);
        vue.nextTick(() => {
          if (inputRef.value && computedValue.value !== inputRef.value.value) {
            inputRef.value.value = computedValue.value;
          }
        });
      }
    };
    const handleClear = (ev) => {
      updateValue("");
      emit("clear", ev);
    };
    const handleKeyDown = (e) => {
      const keyCode = e.key || e.code;
      if (!isComposition.value && keyCode === keycode.Enter.key) {
        preValue = computedValue.value;
        emit("change", computedValue.value, e);
        emit("pressEnter", e);
      }
    };
    const outerCls = vue.computed(() => [`${prefixCls}-outer`, `${prefixCls}-outer-size-${props.size}`, {
      [`${prefixCls}-outer-has-suffix`]: Boolean(slots.suffix),
      [`${prefixCls}-outer-disabled`]: props.disabled
    }]);
    const wrapperCls = vue.computed(() => [`${prefixCls}-wrapper`, {
      [`${prefixCls}-error`]: mergedError.value,
      [`${prefixCls}-disabled`]: props.disabled,
      [`${prefixCls}-focus`]: focused.value
    }]);
    const cls = vue.computed(() => [prefixCls, `${prefixCls}-size-${props.size}`]);
    const wrapperAttrs = vue.computed(() => omit.omit(attrs, constant.INPUT_EVENTS));
    const inputAttrs = vue.computed(() => pick(attrs, constant.INPUT_EVENTS));
    const renderInput = (hasOuter) => {
      var _a;
      return vue.createVNode("span", vue.mergeProps({
        "class": wrapperCls.value,
        "onMousedown": handleMousedown
      }, !hasOuter ? wrapperAttrs.value : void 0), [slots.prefix && vue.createVNode("span", {
        "class": `${prefixCls}-prefix`
      }, [slots.prefix()]), vue.createVNode("input", vue.mergeProps(inputAttrs.value, {
        "ref": inputRef,
        "class": cls.value,
        "value": computedValue.value,
        "type": props.type,
        "placeholder": props.placeholder,
        "readonly": props.readonly,
        "disabled": props.disabled,
        "onInput": handleInput,
        "onKeydown": handleKeyDown,
        "onFocus": handleFocus,
        "onBlur": handleBlur,
        "onCompositionstart": handleComposition,
        "onCompositionupdate": handleComposition,
        "onCompositionend": handleComposition
      }), null), showClearBtn.value && vue.createVNode(iconHover, {
        "prefix": prefixCls,
        "class": `${prefixCls}-clear-btn`,
        "onClick": handleClear
      }, {
        default: () => [vue.createVNode(index, null, null)]
      }), (slots.suffix || Boolean(props.maxLength) && props.showWordLimit) && vue.createVNode("span", {
        "class": `${prefixCls}-suffix`
      }, [Boolean(props.maxLength) && props.showWordLimit && vue.createVNode("span", {
        "class": `${prefixCls}-word-limit`
      }, [valueLength.value, vue.createTextVNode("/"), maxLength.value]), (_a = slots.suffix) == null ? void 0 : _a.call(slots)])]);
    };
    const render = () => {
      if (slots.prepend || slots.append) {
        return vue.createVNode("span", vue.mergeProps({
          "class": outerCls.value
        }, wrapperAttrs.value), [slots.prepend && vue.createVNode("span", {
          "class": `${prefixCls}-prepend`
        }, [slots.prepend()]), renderInput(true), slots.append && vue.createVNode("span", {
          "class": `${prefixCls}-append`
        }, [slots.append()])]);
      }
      return renderInput();
    };
    return {
      inputRef,
      render
    };
  },
  methods: {
    focus() {
      var _a;
      (_a = this.inputRef) == null ? void 0 : _a.focus();
    },
    blur() {
      var _a;
      (_a = this.inputRef) == null ? void 0 : _a.blur();
    }
  },
  render() {
    return this.render();
  }
});
module.exports = _Input;
