// @arco-design/web-vue/lib/menu/hooks/use-menu-data-collector.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports[Symbol.toStringTag] = "Module";
var vue = require("vue");
var context = require("../context.js");
function getKeys(data, condition) {
  const keys = [];
  const loop = (list) => {
    list.forEach((item) => {
      if (condition(item)) {
        keys.push(item.key);
      }
      if (item.children) {
        loop(item.children);
      }
    });
  };
  loop(data);
  return keys;
}
function useMenuDataCollectorProvider() {
  const data = vue.ref([]);
  const subMenuKeys = vue.computed(() => getKeys(data.value, (item) => !!item.children));
  const menuItemKeys = vue.computed(() => getKeys(data.value, (item) => !item.children));
  vue.provide(context.DataCollectorInjectionKey, {
    collectSubMenu(key, children) {
      data.value = [
        ...data.value,
        {
          key,
          children
        }
      ];
    },
    removeSubMenu(key) {
      data.value = data.value.filter((item) => item.key !== key);
    },
    collectMenuItem(key) {
      data.value.push({ key });
    },
    removeMenuItem(key) {
      data.value = data.value.filter((item) => item.key !== key);
    },
    reportMenuData(reportData) {
      data.value = reportData;
    }
  });
  return {
    data,
    subMenuKeys,
    menuItemKeys
  };
}
function useMenuDataCollectorContext(isRoot = false) {
  const menuContext = isRoot ? {} : vue.inject(context.DataCollectorInjectionKey);
  return menuContext || {};
}
function useMenuDataCollector(props = { isRoot: false }) {
  const { isRoot, key } = props;
  const { data, subMenuKeys, menuItemKeys } = useMenuDataCollectorProvider();
  const { collectSubMenu, removeSubMenu, reportMenuData } = useMenuDataCollectorContext(isRoot);
  if (key !== void 0) {
    vue.onMounted(() => {
      collectSubMenu && collectSubMenu(key, data.value);
    });
    vue.onUnmounted(() => {
      removeSubMenu && removeSubMenu(key);
    });
  }
  if (!isRoot && key === void 0) {
    vue.onMounted(() => {
      reportMenuData && reportMenuData(data.value);
    });
  }
  return {
    menuData: data,
    subMenuKeys,
    menuItemKeys
  };
}
exports["default"] = useMenuDataCollector;
exports.useMenuDataCollectorContext = useMenuDataCollectorContext;
exports.useMenuDataCollectorProvider = useMenuDataCollectorProvider;
