// @arco-design/web-vue/lib/_utils/constant.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports[Symbol.toStringTag] = "Module";
const SIZES = ["mini", "small", "medium", "large"];
const BORDER_SHAPES = ["square", "round", "circle"];
const STATUSES = ["normal", "success", "warning", "danger"];
const MESSAGE_TYPES = ["info", "success", "warning", "error"];
const TRIGGER_POSITIONS = [
  "top",
  "tl",
  "tr",
  "bottom",
  "bl",
  "br",
  "left",
  "lt",
  "lb",
  "right",
  "rt",
  "rb"
];
const TRIGGER_EVENTS = [
  "hover",
  "click",
  "focus",
  "contextMenu"
];
const DIRECTIONS = ["horizontal", "vertical"];
const INPUT_EVENTS = [
  "onFocus",
  "onFocusin",
  "onFocusout",
  "onBlur",
  "onChange",
  "onBeforeinput",
  "onInput",
  "onReset",
  "onSubmit",
  "onInvalid",
  "onKeydown",
  "onKeypress",
  "onKeyup",
  "onCopy",
  "onCut",
  "onPaste",
  "onCompositionstart",
  "onCompositionupdate",
  "onCompositionend",
  "onSelect"
];
exports.BORDER_SHAPES = BORDER_SHAPES;
exports.DIRECTIONS = DIRECTIONS;
exports.INPUT_EVENTS = INPUT_EVENTS;
exports.MESSAGE_TYPES = MESSAGE_TYPES;
exports.SIZES = SIZES;
exports.STATUSES = STATUSES;
exports.TRIGGER_EVENTS = TRIGGER_EVENTS;
exports.TRIGGER_POSITIONS = TRIGGER_POSITIONS;
