// @arco-design/web-vue/lib/divider/divider.js
"use strict";
var vue = require("vue");
var globalConfig = require("../_utils/global-config.js");
var _Divider = vue.defineComponent({
  name: "Divider",
  props: {
    direction: {
      type: String,
      validator: (value) => ["horizontal", "vertical"].includes(value),
      default: "horizontal"
    },
    orientation: {
      type: String,
      validator: (value) => ["left", "right", "center"].includes(value),
      default: "center"
    }
  },
  setup(props, {
    slots
  }) {
    const prefixCls = globalConfig.getPrefixCls("divider");
    return () => {
      var _a;
      const {
        direction,
        orientation
      } = props;
      const children = (_a = slots.default) == null ? void 0 : _a.call(slots);
      const classNames = [prefixCls, `${prefixCls}-${direction}`, {
        [`${prefixCls}-with-text`]: children
      }];
      return vue.createVNode("div", {
        "class": classNames
      }, [children && direction === "horizontal" ? vue.createVNode("span", {
        "class": `${prefixCls}-text ${prefixCls}-text-${orientation}`
      }, [children]) : null]);
    };
  }
});
module.exports = _Divider;
