// @arco-design/web-vue/es/_components/select-view/select-view.js
import { defineComponent, toRefs, ref, computed, watch, createVNode, Fragment } from "vue";
import { getPrefixCls } from "../../_utils/global-config.js";
import { isArray } from "../../_utils/is.js";
import InputLabel from "../input-label/input-label.js";
import InputTag from "../../input-tag/index.js";
import IconHover from "../icon-hover.js";
import IconDown from "../../icon/icon-down/index.js";
import IconLoading from "../../icon/icon-loading/index.js";
import IconClose from "../../icon/icon-close/index.js";
import IconExpand from "../../icon/icon-expand/index.js";
import IconSearch from "../../icon/icon-search/index.js";
var SelectView = defineComponent({
  name: "SelectView",
  props: {
    modelValue: {
      type: [Object, Array]
    },
    inputValue: String,
    placeholder: String,
    disabled: {
      type: Boolean,
      default: false
    },
    error: {
      type: Boolean,
      default: false
    },
    loading: {
      type: Boolean,
      default: false
    },
    opened: {
      type: Boolean,
      default: false
    },
    size: {
      type: String,
      default: "medium"
    },
    bordered: {
      type: Boolean,
      default: true
    },
    multiple: {
      type: Boolean,
      default: false
    },
    allowClear: {
      type: Boolean,
      default: false
    },
    allowCreate: {
      type: Boolean,
      default: false
    },
    allowSearch: {
      type: Boolean,
      default: (props) => isArray(props.modelValue)
    },
    maxTagCount: {
      type: Number,
      default: 0
    },
    formatLabel: {
      type: Function
    },
    retainInputValue: {
      type: Boolean,
      default: false
    }
  },
  emits: ["remove", "clear"],
  setup(props, {
    emit,
    slots,
    attrs
  }) {
    const prefixCls = getPrefixCls("select-view");
    const {
      opened
    } = toRefs(props);
    const componentRef = ref();
    const inputRef = computed(() => {
      var _a;
      return (_a = componentRef.value) == null ? void 0 : _a.inputRef;
    });
    const validValue = computed(() => {
      if (props.multiple && !isArray(props.modelValue)) {
        return props.modelValue ? [props.modelValue] : [];
      }
      if (!props.multiple && isArray(props.modelValue)) {
        return props.modelValue[0];
      }
      return props.modelValue;
    });
    const isEmptyValue = computed(() => isArray(validValue.value) ? validValue.value.length === 0 : !validValue.value);
    const enabledInput = computed(() => props.allowSearch || props.allowCreate);
    const showClearBtn = computed(() => props.allowClear && !props.disabled && !isEmptyValue.value);
    const handleRemove = (tag) => {
      emit("remove", tag);
    };
    const handleClear = (ev) => {
      emit("clear", ev);
    };
    const renderIcon = () => {
      var _a, _b, _c, _d;
      if (props.loading) {
        return (_b = (_a = slots.loadingIcon) == null ? void 0 : _a.call(slots)) != null ? _b : createVNode(IconLoading, null, null);
      }
      if (props.allowSearch && props.opened) {
        return (_d = (_c = slots.searchIcon) == null ? void 0 : _c.call(slots)) != null ? _d : createVNode(IconSearch, null, null);
      }
      if (slots.arrowIcon) {
        return slots.arrowIcon();
      }
      if (props.multiple || enabledInput.value) {
        return createVNode(IconExpand, {
          "style": {
            transform: "rotate(-45deg)"
          }
        }, null);
      }
      return createVNode(IconDown, {
        "class": `${prefixCls}-arrow-icon`
      }, null);
    };
    const renderSuffix = () => createVNode(Fragment, null, [showClearBtn.value && createVNode(IconHover, {
      "class": `${prefixCls}-clear-btn`,
      "onClick": handleClear,
      "onMousedown": (e) => e.stopPropagation()
    }, {
      default: () => [createVNode(IconClose, null, null)]
    }), createVNode("span", {
      "class": `${prefixCls}-icon`
    }, [renderIcon()])]);
    watch(opened, (opened2) => {
      if (!opened2 && inputRef.value && inputRef.value.isSameNode(document.activeElement)) {
        inputRef.value.blur();
      }
    });
    const cls = computed(() => [`${prefixCls}-${props.multiple ? "multiple" : "single"}`, {
      [`${prefixCls}-opened`]: props.opened,
      [`${prefixCls}-borderless`]: !props.bordered
    }]);
    const render = () => {
      if (isArray(validValue.value)) {
        return createVNode(InputTag, {
          "ref": componentRef,
          "baseCls": prefixCls,
          "class": cls.value,
          "modelValue": validValue.value,
          "inputValue": props.inputValue,
          "formatTag": props.formatLabel,
          "focused": props.opened,
          "placeholder": props.placeholder,
          "disabled": props.disabled,
          "size": props.size,
          "error": props.error,
          "maxTagCount": props.maxTagCount,
          "disabledInput": !props.allowSearch && !props.allowCreate,
          "retainInputValue": true,
          "onRemove": handleRemove
        }, {
          prefix: slots.prefix,
          suffix: renderSuffix,
          tag: slots.label
        });
      }
      return createVNode(InputLabel, {
        "ref": componentRef,
        "baseCls": prefixCls,
        "class": cls.value,
        "modelValue": validValue.value,
        "inputValue": props.inputValue,
        "focused": props.opened,
        "placeholder": props.placeholder,
        "disabled": props.disabled,
        "size": props.size,
        "error": props.error,
        "formatLabel": props.formatLabel,
        "enabledInput": enabledInput.value
      }, {
        default: slots.label,
        prefix: slots.prefix,
        suffix: renderSuffix
      });
    };
    return {
      inputRef,
      render
    };
  },
  methods: {
    focus() {
      if (this.inputRef) {
        this.inputRef.focus();
      }
    },
    blur() {
      if (this.inputRef) {
        this.inputRef.blur();
      }
    }
  },
  render() {
    return this.render();
  }
});
export { SelectView as default };
