// @arco-design/web-vue/lib/modal/index.js
"use strict";
var __defProp = Object.defineProperty;
var __defProps = Object.defineProperties;
var __getOwnPropDescs = Object.getOwnPropertyDescriptors;
var __getOwnPropSymbols = Object.getOwnPropertySymbols;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __propIsEnum = Object.prototype.propertyIsEnumerable;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __spreadValues = (a, b) => {
  for (var prop in b || (b = {}))
    if (__hasOwnProp.call(b, prop))
      __defNormalProp(a, prop, b[prop]);
  if (__getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(b)) {
      if (__propIsEnum.call(b, prop))
        __defNormalProp(a, prop, b[prop]);
    }
  return a;
};
var __spreadProps = (a, b) => __defProps(a, __getOwnPropDescs(b));
var vue = require("vue");
var globalConfig = require("../_utils/global-config.js");
var constant = require("../_utils/constant.js");
var dom = require("../_utils/dom.js");
var is = require("../_utils/is.js");
var modal$1 = require("./modal.js");
var omit = require("../_utils/omit.js");
const open = (config) => {
  const container = dom.getOverlay("modal");
  const handleOk = () => {
    if (vm.component) {
      vm.component.props.visible = false;
    }
    if (is.isFunction(config.onOk)) {
      config.onOk();
    }
  };
  const handleCancel = () => {
    if (vm.component) {
      vm.component.props.visible = false;
    }
    if (is.isFunction(config.onCancel)) {
      config.onCancel();
    }
  };
  const handleClose = () => {
    vue.render(null, container);
    document.body.removeChild(container);
    if (is.isFunction(config.onClose)) {
      config.onClose();
    }
  };
  const defaultConfig = {
    visible: true,
    renderToBody: false,
    onOk: handleOk,
    onCancel: handleCancel,
    onClose: handleClose
  };
  const vm = vue.createVNode(modal$1, __spreadValues(__spreadValues({}, omit.omit(config, ["content", "title", "footer"])), defaultConfig), {
    default: is.isFunction(config.content) ? config.content : () => config.content,
    title: is.isFunction(config.title) ? config.title : () => config.title,
    footer: is.isFunction(config.footer) ? config.footer : () => config.footer
  });
  vue.render(vm, container);
  document.body.appendChild(container);
  return {
    close
  };
};
const modal = __spreadValues({
  open,
  confirm: (config) => {
    const _config = __spreadValues({
      simple: true
    }, config);
    return open(_config);
  }
}, constant.MESSAGE_TYPES.reduce((pre, value) => {
  pre[value] = (config) => {
    const _config = __spreadValues({
      simple: true,
      hideCancel: true,
      messageType: value
    }, config);
    return open(_config);
  };
  return pre;
}, {}));
const Modal = Object.assign(modal$1, __spreadProps(__spreadValues({}, modal), {
  install: (app, options) => {
    globalConfig.setGlobalConfig(app, options);
    const componentPrefix = globalConfig.getComponentPrefix(options);
    app.component(componentPrefix + modal$1.name, modal$1);
    app.config.globalProperties.$modal = modal;
  }
}));
module.exports = Modal;
