// @arco-design/web-vue/es/tree/hooks/use-checked-state.js
import { toRefs, ref, watchEffect, computed } from "vue";
import { getCheckedStateByInitKeys } from "../utils/index.js";
function useCheckedState(props) {
  const {
    defaultCheckedKeys,
    checkedKeys: propCheckedKeys,
    key2TreeNode,
    checkStrictly
  } = toRefs(props);
  const getStateByInitKeys = (initKeys) => {
    return getCheckedStateByInitKeys({
      initCheckedKeys: initKeys,
      key2TreeNode: key2TreeNode.value,
      checkStrictly: checkStrictly.value
    });
  };
  const isInitialized = ref(false);
  const initLocalState = getStateByInitKeys(propCheckedKeys.value || (defaultCheckedKeys == null ? void 0 : defaultCheckedKeys.value) || []);
  const localCheckedKeys = ref(initLocalState[0]);
  const localIndeterminateKeys = ref(initLocalState[1]);
  const computedCheckedKeys = ref();
  const computedIndeterminateKeys = ref();
  watchEffect(() => {
    if (propCheckedKeys.value) {
      [computedCheckedKeys.value, computedIndeterminateKeys.value] = getStateByInitKeys(propCheckedKeys.value);
    } else if (isInitialized.value) {
      computedCheckedKeys.value = void 0;
      computedIndeterminateKeys.value = void 0;
      localCheckedKeys.value = [];
      localIndeterminateKeys.value = [];
    }
    if (!isInitialized.value) {
      isInitialized.value = true;
    }
  });
  return {
    checkedKeys: computed(() => computedCheckedKeys.value || localCheckedKeys.value),
    indeterminateKeys: computed(() => computedIndeterminateKeys.value || localIndeterminateKeys.value),
    setCheckedState(newCheckedKeys, newIndeterminateKeys) {
      localCheckedKeys.value = newCheckedKeys;
      localIndeterminateKeys.value = newIndeterminateKeys;
    }
  };
}
export { useCheckedState as default };
