// @arco-design/web-vue/es/grid/base-grid.js
import { defineComponent, toRefs, computed, createVNode } from "vue";
import { getPrefixCls } from "../_utils/global-config.js";
var BaseGrid = defineComponent({
  name: "BaseGrid",
  props: {
    cols: {
      type: Number,
      default: 24
    },
    rowGap: {
      type: Number,
      default: 0
    },
    colGap: {
      type: Number,
      default: 0
    },
    collapsed: {
      type: Boolean,
      default: false
    },
    collapsedRows: {
      type: Number,
      default: 1
    }
  },
  setup(props, {
    slots
  }) {
    const {
      cols,
      colGap,
      rowGap
    } = toRefs(props);
    const prefixCls = getPrefixCls("grid");
    const classNames = computed(() => [prefixCls]);
    const style = computed(() => [{
      "gap": `${rowGap.value}px ${colGap.value}px`,
      "grid-template-columns": `repeat(${cols.value}, minmax(0px, 1fr))`
    }]);
    return () => {
      var _a;
      return createVNode("div", {
        "class": classNames.value,
        "style": style.value
      }, [(_a = slots.default) == null ? void 0 : _a.call(slots)]);
    };
  }
});
export { BaseGrid as default };
