// @arco-design/web-vue/lib/table/table-tr.js
"use strict";
var vue = require("vue");
var globalConfig = require("../_utils/global-config.js");
var pluginVue_exportHelper = require("../_virtual/plugin-vue_export-helper");
const _sfc_main = vue.defineComponent({
  name: "Tr",
  props: {
    isExpandRow: {
      type: Boolean
    },
    isEmptyRow: {
      type: Boolean
    },
    checked: {
      type: Boolean
    }
  },
  setup(props) {
    const prefixCls = globalConfig.getPrefixCls("table");
    const cls = vue.computed(() => [
      `${prefixCls}-tr`,
      {
        [`${prefixCls}-tr-expand`]: props.isExpandRow,
        [`${prefixCls}-tr-empty`]: props.isEmptyRow,
        [`${prefixCls}-tr-checked`]: props.checked
      }
    ]);
    return {
      prefixCls,
      cls
    };
  }
});
function _sfc_render(_ctx, _cache, $props, $setup, $data, $options) {
  return vue.openBlock(), vue.createElementBlock("tr", {
    class: vue.normalizeClass(_ctx.cls)
  }, [
    vue.renderSlot(_ctx.$slots, "default")
  ], 2);
}
var Tr = /* @__PURE__ */ pluginVue_exportHelper(_sfc_main, [["render", _sfc_render]]);
module.exports = Tr;
