// @arco-design/web-vue/lib/table/table-th.js
"use strict";
var vue = require("vue");
var globalConfig = require("../_utils/global-config.js");
var index$5 = require("../checkbox/index.js");
var index$6 = require("../radio/index.js");
var index$7 = require("../button/index.js");
var index$3 = require("../trigger/index.js");
var index$2 = require("../icon/icon-caret-down/index.js");
var index$1 = require("../icon/icon-caret-up/index.js");
var index$4 = require("../icon/icon-filter/index.js");
var useColumnSorter = require("./hooks/use-column-sorter.js");
var useColumnFilter = require("./hooks/use-column-filter.js");
var index = require("../locale/index.js");
var utils = require("./utils.js");
var is = require("../_utils/is.js");
var iconHover = require("../_components/icon-hover.js");
function _isSlot(s) {
  return typeof s === "function" || Object.prototype.toString.call(s) === "[object Object]" && !vue.isVNode(s);
}
var Th = vue.defineComponent({
  name: "Th",
  props: {
    column: {
      type: Object,
      default: () => ({})
    },
    operations: {
      type: Array,
      default: () => []
    },
    dataColumns: {
      type: Array,
      default: () => []
    },
    sortOrder: {
      type: String
    },
    filterValue: {
      type: Array
    },
    filterIconAlignLeft: {
      type: Boolean,
      default: false
    }
  },
  emits: ["sorterChange", "filterChange"],
  setup(props, {
    emit,
    slots
  }) {
    const prefixCls = globalConfig.getPrefixCls("table");
    const {
      t
    } = index.useI18n();
    const filterIconAlignLeft = vue.computed(() => {
      var _a, _b, _c;
      if (is.isBoolean((_b = (_a = props.column) == null ? void 0 : _a.filterable) == null ? void 0 : _b.alignLeft)) {
        return (_c = props.column.filterable) == null ? void 0 : _c.alignLeft;
      }
      return props.filterIconAlignLeft;
    });
    const {
      hasSorter,
      hasAscendBtn,
      hasDescendBtn,
      nextSortOrder,
      handleClickSorter
    } = useColumnSorter.useColumnSorter(props, emit);
    const {
      filterPopupVisible,
      isFilterActive,
      isMultipleFilter,
      columnFilterValue,
      handleFilterPopupVisibleChange,
      setFilterValue,
      handleCheckboxFilterChange,
      handleRadioFilterChange,
      handleFilterConfirm,
      handleFilterReset
    } = useColumnFilter.useColumnFilter(props, emit);
    const renderFilterContent = () => {
      let _slot, _slot2;
      const {
        filterable
      } = props.column;
      if (filterable == null ? void 0 : filterable.renderContent) {
        return filterable.renderContent({
          filterValue: columnFilterValue.value,
          setFilterValue,
          handleFilterConfirm,
          handleFilterReset
        });
      }
      return vue.createVNode("div", {
        "class": `${prefixCls}-filters-content`
      }, [vue.createVNode("ul", {
        "class": `${prefixCls}-filters-list`
      }, [filterable == null ? void 0 : filterable.filters.map((item, index2) => {
        var _a;
        return vue.createVNode("li", {
          "class": `${prefixCls}-filters-item`,
          "key": index2
        }, [isMultipleFilter.value ? vue.createVNode(index$5["default"], {
          "value": item.value,
          "modelValue": columnFilterValue.value,
          "uninjectGroupContext": true,
          "onChange": handleCheckboxFilterChange
        }, {
          default: () => [item.text]
        }) : vue.createVNode(index$6["default"], {
          "value": item.value,
          "modelValue": (_a = columnFilterValue.value[0]) != null ? _a : "",
          "uninjectGroupContext": true,
          "onChange": handleRadioFilterChange
        }, {
          default: () => [item.text]
        })]);
      })]), vue.createVNode("div", {
        "class": `${prefixCls}-filters-bottom`
      }, [vue.createVNode(index$7["default"], {
        "size": "mini",
        "onClick": handleFilterReset
      }, _isSlot(_slot = t("table.resetText")) ? _slot : {
        default: () => [_slot]
      }), vue.createVNode(index$7["default"], {
        "type": "primary",
        "size": "mini",
        "onClick": handleFilterConfirm
      }, _isSlot(_slot2 = t("table.okText")) ? _slot2 : {
        default: () => [_slot2]
      })])]);
    };
    const renderFilter = () => {
      const {
        filterable
      } = props.column;
      if (!filterable || filterable.filters.length === 0) {
        return null;
      }
      return vue.createVNode(index$3, vue.mergeProps({
        "popupVisible": filterPopupVisible.value,
        "trigger": "click",
        "autoFitPosition": true,
        "popupOffset": filterIconAlignLeft.value ? 4 : 0,
        "onPopupVisibleChange": handleFilterPopupVisibleChange
      }, filterable.triggerProps), {
        default: () => [vue.createVNode(iconHover, {
          "class": [`${prefixCls}-filters`, {
            [`${prefixCls}-filters-active`]: isFilterActive.value,
            [`${prefixCls}-filters-open`]: filterPopupVisible.value,
            [`${prefixCls}-filters-align-left`]: filterIconAlignLeft.value
          }],
          "disabled": !filterIconAlignLeft.value,
          "onClick": (ev) => ev.stopPropagation()
        }, {
          default: () => {
            var _a, _b;
            return [(_b = (_a = filterable.icon) == null ? void 0 : _a.call(filterable)) != null ? _b : vue.createVNode(index$4, null, null)];
          }
        })],
        content: renderFilterContent
      });
    };
    const cellCls = vue.computed(() => {
      var _a;
      const cls2 = [`${prefixCls}-cell`, {
        [`${prefixCls}-cell-text-ellipsis`]: (_a = props.column) == null ? void 0 : _a.ellipsis
      }];
      if (hasSorter.value) {
        cls2.push(`${prefixCls}-cell-with-sorter`, {
          [`${prefixCls}-cell-next-ascend`]: nextSortOrder.value === "ascend",
          [`${prefixCls}-cell-next-descend`]: nextSortOrder.value === "descend"
        });
      }
      if (filterIconAlignLeft.value) {
        cls2.push(`${prefixCls}-cell-with-filter`);
      }
      return cls2;
    });
    const renderTitle = () => {
      if (slots.default) {
        return slots.default();
      }
      if (is.isFunction(props.column.title)) {
        return props.column.title();
      }
      return props.column.title;
    };
    const renderCell = () => {
      var _a;
      return vue.createVNode("span", {
        "class": cellCls.value,
        "style": (_a = props.column) == null ? void 0 : _a.cellStyle,
        "onClick": hasSorter.value ? handleClickSorter : void 0
      }, [vue.createVNode("span", {
        "class": `${prefixCls}-th-item-title`
      }, [renderTitle()]), hasSorter.value && vue.createVNode("span", {
        "class": `${prefixCls}-sorter`
      }, [hasAscendBtn.value && vue.createVNode("div", {
        "class": [`${prefixCls}-sorter-icon`, {
          [`${prefixCls}-sorter-icon-active`]: props.sortOrder === "ascend"
        }]
      }, [vue.createVNode(index$1, null, null)]), hasDescendBtn.value && vue.createVNode("div", {
        "class": [`${prefixCls}-sorter-icon`, {
          [`${prefixCls}-sorter-icon-active`]: props.sortOrder === "descend"
        }]
      }, [vue.createVNode(index$2, null, null)])]), filterIconAlignLeft.value && renderFilter()]);
    };
    const style = vue.computed(() => utils.getStyle(props.column, {
      dataColumns: props.dataColumns,
      operations: props.operations
    }));
    const cls = vue.computed(() => {
      var _a, _b;
      return [`${prefixCls}-th`, `${prefixCls}-th-align-${(_b = (_a = props.column) == null ? void 0 : _a.align) != null ? _b : "left"}`, {
        [`${prefixCls}-col-sorted`]: Boolean(props.sortOrder)
      }, ...utils.getFixedCls(prefixCls, props.column)];
    });
    return () => {
      var _a, _b;
      const colSpan = (_a = props.column.colSpan) != null ? _a : 1;
      const rowSpan = (_b = props.column.rowSpan) != null ? _b : 1;
      return vue.createVNode("th", {
        "class": cls.value,
        "style": style.value,
        "colspan": colSpan > 1 ? colSpan : void 0,
        "rowspan": rowSpan > 1 ? rowSpan : void 0
      }, [renderCell(), !filterIconAlignLeft.value && renderFilter()]);
    };
  }
});
module.exports = Th;
