// @arco-design/web-vue/lib/_components/input-label/input-label.js
"use strict";
var vue = require("vue");
var omit = require("../../_utils/omit.js");
var constant = require("../../_utils/constant.js");
var pick = require("../../_utils/pick.js");
var globalConfig = require("../../_utils/global-config.js");
var useInput = require("../../_hooks/use-input.js");
var InputLabel = vue.defineComponent({
  name: "InputLabel",
  inheritAttrs: false,
  props: {
    modelValue: Object,
    inputValue: String,
    enabledInput: Boolean,
    formatLabel: Function,
    placeholder: String,
    retainInputValue: Boolean,
    disabled: Boolean,
    baseCls: String,
    size: String,
    error: Boolean,
    focused: Boolean
  },
  emits: ["update:inputValue", "inputValueChange", "focus", "blur"],
  setup(props, {
    attrs,
    emit,
    slots
  }) {
    var _a;
    const prefixCls = (_a = props.baseCls) != null ? _a : globalConfig.getPrefixCls("input-label");
    const {
      inputRef,
      _focused,
      computedValue: computedInputValue,
      handleInput,
      handleComposition,
      handleFocus,
      handleBlur,
      handleMousedown
    } = useInput.useInput(props, {
      emit,
      isInputValue: true
    });
    const mergedFocused = vue.computed(() => {
      var _a2;
      return (_a2 = props.focused) != null ? _a2 : _focused.value;
    });
    const showInput = vue.computed(() => props.enabledInput && _focused.value || !props.modelValue);
    const mergedPlaceholder = vue.computed(() => {
      if (props.enabledInput && props.modelValue) {
        return props.modelValue.label;
      }
      return props.placeholder;
    });
    const cls = vue.computed(() => [prefixCls, `${prefixCls}-size-${props.size}`, {
      [`${prefixCls}-search`]: props.enabledInput,
      [`${prefixCls}-focus`]: mergedFocused.value,
      [`${prefixCls}-disabled`]: props.disabled,
      [`${prefixCls}-error`]: props.error
    }]);
    const wrapperAttrs = vue.computed(() => omit.omit(attrs, constant.INPUT_EVENTS));
    const inputAttrs = vue.computed(() => pick(attrs, constant.INPUT_EVENTS));
    const render = () => {
      var _a2, _b, _c, _d, _e;
      return vue.createVNode("span", vue.mergeProps(wrapperAttrs.value, {
        "class": cls.value,
        "onMousedown": handleMousedown
      }), [slots.prefix && vue.createVNode("span", {
        "class": `${prefixCls}-prefix`
      }, [slots.prefix()]), vue.createVNode("input", vue.mergeProps(inputAttrs.value, {
        "ref": inputRef,
        "class": [`${prefixCls}-input`, {
          [`${prefixCls}-input-hidden`]: !showInput.value
        }],
        "value": computedInputValue.value,
        "readonly": !props.enabledInput,
        "placeholder": mergedPlaceholder.value,
        "disabled": props.disabled,
        "onInput": handleInput,
        "onFocus": handleFocus,
        "onBlur": handleBlur,
        "onCompositionstart": handleComposition,
        "onCompositionupdate": handleComposition,
        "onCompositionend": handleComposition
      }), null), vue.createVNode("span", {
        "class": [`${prefixCls}-value`, {
          [`${prefixCls}-value-hidden`]: showInput.value
        }]
      }, [(_e = (_c = (_a2 = slots.default) == null ? void 0 : _a2.call(slots, {
        data: props.modelValue
      })) != null ? _c : (_b = props.formatLabel) == null ? void 0 : _b.call(props, props.modelValue)) != null ? _e : (_d = props.modelValue) == null ? void 0 : _d.label]), slots.suffix && vue.createVNode("span", {
        "class": `${prefixCls}-suffix`
      }, [slots.suffix()])]);
    };
    return {
      inputRef,
      render
    };
  },
  methods: {
    focus() {
      var _a;
      (_a = this.inputRef) == null ? void 0 : _a.focus();
    },
    blur() {
      var _a;
      (_a = this.inputRef) == null ? void 0 : _a.blur();
    }
  },
  render() {
    return this.render();
  }
});
module.exports = InputLabel;
