// @arco-design/web-vue/es/transfer/transfer-list-item.js
import { defineComponent, inject, computed, createVNode } from "vue";
import { getPrefixCls } from "../_utils/global-config.js";
import IconHover from "../_components/icon-hover.js";
import Checkbox from "../checkbox/index.js";
import IconClose from "../icon/icon-close/index.js";
import { transferInjectionKey } from "./context.js";
var TransferListItem = defineComponent({
  name: "TransferListItem",
  props: {
    type: {
      type: String
    },
    data: {
      type: Object,
      required: true
    },
    allowClear: {
      type: Boolean
    },
    disabled: {
      type: Boolean
    },
    onSelectChange: {
      type: Function
    },
    draggable: {
      type: Boolean
    },
    simple: Boolean
  },
  setup(props) {
    const prefixCls = getPrefixCls("transfer-list-item");
    const transferCtx = inject(transferInjectionKey, void 0);
    const handleClick = () => {
      if (!props.simple) {
        return;
      }
      transferCtx == null ? void 0 : transferCtx.moveTo([props.data.value], props.type === "target" ? "source" : "target");
    };
    const cls = computed(() => [prefixCls, {
      [`${prefixCls}-disabled`]: props.disabled,
      [`${prefixCls}-draggable`]: props.draggable
    }]);
    const handleRemove = () => {
      transferCtx == null ? void 0 : transferCtx.moveTo([props.data.value], "source");
    };
    return () => {
      var _a, _b;
      return createVNode("div", {
        "class": cls.value,
        "onClick": handleClick
      }, [props.allowClear || props.simple ? createVNode("span", {
        "class": `${prefixCls}-content`
      }, [(_b = (_a = transferCtx == null ? void 0 : transferCtx.itemSlot) == null ? void 0 : _a.call(transferCtx, {
        label: props.data.label
      })) != null ? _b : props.data.label]) : createVNode(Checkbox, {
        "class": `${prefixCls}-content`,
        "modelValue": transferCtx == null ? void 0 : transferCtx.selected,
        "onChange": props.onSelectChange,
        "value": props.data.value,
        "uninjectGroupContext": true
      }, {
        default: () => {
          var _a2, _b2;
          return [(_b2 = (_a2 = transferCtx == null ? void 0 : transferCtx.itemSlot) == null ? void 0 : _a2.call(transferCtx, {
            label: props.data.label
          })) != null ? _b2 : props.data.label];
        }
      }), props.allowClear && !props.disabled && createVNode(IconHover, {
        "class": `${prefixCls}-remove-btn`,
        "onClick": handleRemove
      }, {
        default: () => [createVNode(IconClose, null, null)]
      })]);
    };
  }
});
export { TransferListItem as default };
