// @arco-design/web-vue/es/button/button.js
import { defineComponent, inject, computed, resolveComponent, openBlock, createElementBlock, normalizeClass, createBlock, renderSlot, createCommentVNode } from "vue";
import { BORDER_SHAPES } from "../_utils/constant.js";
import { getPrefixCls } from "../_utils/global-config.js";
import { isString } from "../_utils/is.js";
import IconLoading from "../icon/icon-loading/index.js";
import { configProviderInjectionKey } from "../config-provider/context.js";
import _export_sfc from "../_virtual/plugin-vue_export-helper";
const BUTTON_TYPES = [
  "primary",
  "secondary",
  "outline",
  "dashed",
  "text"
];
const _sfc_main = defineComponent({
  name: "Button",
  components: {
    IconLoading
  },
  props: {
    type: {
      type: String,
      default: "secondary",
      validator: (value) => {
        return BUTTON_TYPES.includes(value);
      }
    },
    shape: {
      type: String,
      default: "square",
      validator: (value) => {
        return BORDER_SHAPES.includes(value);
      }
    },
    status: {
      type: String,
      default: "normal"
    },
    size: {
      type: String,
      default: () => {
        var _a, _b;
        return (_b = (_a = inject(configProviderInjectionKey, void 0)) == null ? void 0 : _a.size) != null ? _b : "medium";
      }
    },
    long: {
      type: Boolean,
      default: false
    },
    loading: {
      type: Boolean,
      default: false
    },
    disabled: {
      type: Boolean,
      default: false
    },
    htmlType: {
      type: String,
      default: "button"
    },
    href: String,
    onClick: {
      type: [Function, Array]
    }
  },
  emits: [
    "click"
  ],
  setup(props, { emit }) {
    const prefixCls = getPrefixCls("btn");
    const cls = computed(() => [
      prefixCls,
      `${prefixCls}-${props.type}`,
      `${prefixCls}-shape-${props.shape}`,
      `${prefixCls}-size-${props.size}`,
      `${prefixCls}-status-${props.status}`,
      {
        [`${prefixCls}-long`]: props.long,
        [`${prefixCls}-loading`]: props.loading,
        [`${prefixCls}-disabled`]: props.disabled,
        [`${prefixCls}-link`]: isString(props.href)
      }
    ]);
    const handleClick = (ev) => {
      if (props.disabled || props.loading) {
        return;
      }
      emit("click", ev);
    };
    return {
      prefixCls,
      cls,
      handleClick
    };
  }
});
const _hoisted_1 = ["href"];
const _hoisted_2 = ["type", "disabled"];
function _sfc_render(_ctx, _cache, $props, $setup, $data, $options) {
  const _component_icon_loading = resolveComponent("icon-loading");
  return _ctx.href ? (openBlock(), createElementBlock("a", {
    key: 0,
    class: normalizeClass([
      _ctx.cls,
      { [`${_ctx.prefixCls}-only-icon`]: _ctx.$slots.icon && !_ctx.$slots.default }
    ]),
    href: _ctx.disabled || _ctx.loading ? void 0 : _ctx.href,
    onClick: _cache[0] || (_cache[0] = (...args) => _ctx.handleClick && _ctx.handleClick(...args))
  }, [
    _ctx.loading || _ctx.$slots.icon ? (openBlock(), createElementBlock("span", {
      key: 0,
      class: normalizeClass(`${_ctx.prefixCls}-icon`)
    }, [
      _ctx.loading ? (openBlock(), createBlock(_component_icon_loading, {
        key: 0,
        spin: "true"
      })) : renderSlot(_ctx.$slots, "icon", { key: 1 })
    ], 2)) : createCommentVNode("v-if", true),
    renderSlot(_ctx.$slots, "default")
  ], 10, _hoisted_1)) : (openBlock(), createElementBlock("button", {
    key: 1,
    class: normalizeClass([
      _ctx.cls,
      { [`${_ctx.prefixCls}-only-icon`]: _ctx.$slots.icon && !_ctx.$slots.default }
    ]),
    type: _ctx.htmlType,
    disabled: _ctx.disabled,
    onClick: _cache[1] || (_cache[1] = (...args) => _ctx.handleClick && _ctx.handleClick(...args))
  }, [
    _ctx.loading || _ctx.$slots.icon ? (openBlock(), createElementBlock("span", {
      key: 0,
      class: normalizeClass(`${_ctx.prefixCls}-icon`)
    }, [
      _ctx.loading ? (openBlock(), createBlock(_component_icon_loading, {
        key: 0,
        spin: true
      })) : renderSlot(_ctx.$slots, "icon", { key: 1 })
    ], 2)) : createCommentVNode("v-if", true),
    renderSlot(_ctx.$slots, "default")
  ], 10, _hoisted_2));
}
var _Button = /* @__PURE__ */ _export_sfc(_sfc_main, [["render", _sfc_render]]);
export { _Button as default };
