// @arco-design/web-vue/es/date-picker/hooks/use-header-value.js
import { toRefs, reactive, computed, ref, watch } from "vue";
import { getDayjsValue, getNow, methods } from "../../_utils/date.js";
import usePanelSpan from "./use-panel-span.js";
function useHeaderValue(props) {
  const { mode, value, defaultValue, selectedValue, format, onChange } = toRefs(props);
  const { span, superSpan } = usePanelSpan(reactive({
    mode
  }));
  const isSame = (current, target) => {
    const unit = mode.value === "date" || mode.value === "week" ? "M" : "y";
    return current.isSame(target, unit);
  };
  const computedValue = computed(() => getDayjsValue(value.value, format.value));
  const computedDefaultValue = computed(() => getDayjsValue(defaultValue.value, format.value));
  const localValue = ref(computedDefaultValue.value || getNow());
  const headerValue = computed(() => computedValue.value || localValue.value);
  const setLocalValue = (newVal) => {
    if (!newVal)
      return;
    localValue.value = newVal;
  };
  const setHeaderValue = (newVal, emitChange = true) => {
    var _a;
    if (!newVal)
      return;
    if (emitChange && !isSame(headerValue.value, newVal)) {
      (_a = onChange == null ? void 0 : onChange.value) == null ? void 0 : _a.call(onChange, newVal);
    }
    setLocalValue(newVal);
  };
  if (selectedValue.value) {
    setLocalValue(selectedValue.value);
  }
  watch(selectedValue, (newVal) => {
    setHeaderValue(newVal);
  });
  function getDefaultLocalValue() {
    return selectedValue.value || computedDefaultValue.value || getNow();
  }
  function resetHeaderValue(emitChange = true) {
    const defaultLocalValue = getDefaultLocalValue();
    if (emitChange) {
      setHeaderValue(defaultLocalValue);
    } else {
      setLocalValue(defaultLocalValue);
    }
  }
  const showSingleBtn = computed(() => span.value !== superSpan.value);
  const headerOperations = computed(() => ({
    onSuperPrev: () => {
      setHeaderValue(methods.subtract(headerValue.value, superSpan.value, "M"));
    },
    onPrev: showSingleBtn.value ? () => {
      setHeaderValue(methods.subtract(headerValue.value, span.value, "M"));
    } : void 0,
    onNext: showSingleBtn.value ? () => {
      setHeaderValue(methods.add(headerValue.value, span.value, "M"));
    } : void 0,
    onSuperNext: () => {
      setHeaderValue(methods.add(headerValue.value, superSpan.value, "M"));
    }
  }));
  return [
    headerValue,
    setHeaderValue,
    headerOperations,
    resetHeaderValue,
    getDefaultLocalValue
  ];
}
export { useHeaderValue as default };
