// @arco-design/web-vue/lib/grid/base-grid.js
"use strict";
var vue = require("vue");
var globalConfig = require("../_utils/global-config.js");
var BaseGrid = vue.defineComponent({
  name: "BaseGrid",
  props: {
    cols: {
      type: Number,
      default: 24
    },
    rowGap: {
      type: Number,
      default: 0
    },
    colGap: {
      type: Number,
      default: 0
    },
    collapsed: {
      type: Boolean,
      default: false
    },
    collapsedRows: {
      type: Number,
      default: 1
    }
  },
  setup(props, {
    slots
  }) {
    const {
      cols,
      colGap,
      rowGap
    } = vue.toRefs(props);
    const prefixCls = globalConfig.getPrefixCls("grid");
    const classNames = vue.computed(() => [prefixCls]);
    const style = vue.computed(() => [{
      "gap": `${rowGap.value}px ${colGap.value}px`,
      "grid-template-columns": `repeat(${cols.value}, minmax(0px, 1fr))`
    }]);
    return () => {
      var _a;
      return vue.createVNode("div", {
        "class": classNames.value,
        "style": style.value
      }, [(_a = slots.default) == null ? void 0 : _a.call(slots)]);
    };
  }
});
module.exports = BaseGrid;
