// @arco-design/web-vue/lib/upload/upload-picture-item.js
"use strict";
var vue = require("vue");
var globalConfig = require("../_utils/global-config.js");
var uploadProgress = require("./upload-progress.js");
var context = require("./context.js");
var index = require("../icon/icon-image-close/index.js");
var index$1 = require("../icon/icon-eye/index.js");
var index$2 = require("../icon/icon-upload/index.js");
var index$3 = require("../icon/icon-delete/index.js");
var UploadPictureItem = vue.defineComponent({
  name: "UploadPictureItem",
  props: {
    file: {
      type: Object,
      required: true
    },
    disabled: {
      type: Boolean,
      default: false
    }
  },
  setup(props) {
    const prefixCls = globalConfig.getPrefixCls("upload-list");
    const itemCls = `${prefixCls}-picture`;
    const cls = vue.computed(() => [itemCls, {
      [`${itemCls}-status-error`]: props.file.status === "error"
    }]);
    const uploadCtx = vue.inject(context.uploadInjectionKey, void 0);
    const renderCard = () => {
      var _a, _b, _c, _d, _e, _f, _g, _h;
      if (props.file.status === "uploading") {
        return vue.createVNode(uploadProgress, {
          "file": props.file,
          "listType": "picture-card"
        }, null);
      }
      return vue.createVNode(vue.Fragment, null, [vue.createVNode("img", vue.mergeProps({
        "src": props.file.url,
        "alt": props.file.name
      }, (uploadCtx == null ? void 0 : uploadCtx.imageLoading) ? {
        loading: uploadCtx.imageLoading
      } : void 0), null), vue.createVNode("div", {
        "class": `${itemCls}-mask`
      }, [props.file.status === "error" && (uploadCtx == null ? void 0 : uploadCtx.showCancelButton) && vue.createVNode("div", {
        "class": `${itemCls}-error-tip`
      }, [vue.createVNode("span", {
        "class": [uploadCtx == null ? void 0 : uploadCtx.iconCls, `${uploadCtx == null ? void 0 : uploadCtx.iconCls}-error`]
      }, [((_b = (_a = uploadCtx == null ? void 0 : uploadCtx.customIcon) == null ? void 0 : _a.errorIcon) == null ? void 0 : _b.call(_a)) || vue.createVNode(index, null, null)])]), vue.createVNode("div", {
        "class": `${itemCls}-operation`
      }, [props.file.status !== "error" && vue.createVNode("span", {
        "class": [uploadCtx == null ? void 0 : uploadCtx.iconCls, `${uploadCtx == null ? void 0 : uploadCtx.iconCls}-preview`],
        "onClick": () => uploadCtx == null ? void 0 : uploadCtx.onPreview(props.file)
      }, [((_d = (_c = uploadCtx == null ? void 0 : uploadCtx.customIcon) == null ? void 0 : _c.previewIcon) == null ? void 0 : _d.call(_c)) || vue.createVNode(index$1, null, null)]), ["init", "error"].includes(props.file.status) && (uploadCtx == null ? void 0 : uploadCtx.showRetryButton) && vue.createVNode("span", {
        "class": [uploadCtx == null ? void 0 : uploadCtx.iconCls, `${uploadCtx == null ? void 0 : uploadCtx.iconCls}-upload`],
        "onClick": () => uploadCtx == null ? void 0 : uploadCtx.onUpload(props.file)
      }, [((_f = (_e = uploadCtx == null ? void 0 : uploadCtx.customIcon) == null ? void 0 : _e.retryIcon) == null ? void 0 : _f.call(_e)) || vue.createVNode(index$2, null, null)]), !(uploadCtx == null ? void 0 : uploadCtx.disabled) && (uploadCtx == null ? void 0 : uploadCtx.showRemoveButton) && vue.createVNode("span", {
        "class": [uploadCtx == null ? void 0 : uploadCtx.iconCls, `${uploadCtx == null ? void 0 : uploadCtx.iconCls}-remove`],
        "onClick": () => uploadCtx == null ? void 0 : uploadCtx.onRemove(props.file)
      }, [((_h = (_g = uploadCtx == null ? void 0 : uploadCtx.customIcon) == null ? void 0 : _g.removeIcon) == null ? void 0 : _h.call(_g)) || vue.createVNode(index$3, null, null)])])])]);
    };
    return () => vue.createVNode("span", {
      "class": cls.value
    }, [renderCard()]);
  }
});
module.exports = UploadPictureItem;
