// @arco-design/web-vue/es/dropdown/utils.js
import { isNamedComponent, resolveProps, isSlotsChildren, isArrayChildren, getChildrenString } from "../_utils/vue-utils.js";
import { isArray, isObject } from "../_utils/is.js";
import { omit } from "../_utils/omit.js";
const isGroup = (option) => {
  return isObject(option) && "isGroup" in option;
};
const isSubmenu = (option) => {
  return isObject(option) && "isSubmenu" in option;
};
const travelDropDownChildren = (children) => {
  var _a, _b, _c, _d, _e, _f, _g, _h, _i, _j, _k, _l, _m;
  const options = [];
  for (const child of children) {
    if (!child)
      continue;
    if (isNamedComponent(child, "Dgroup")) {
      const props = resolveProps(child);
      let slots;
      const _options = [];
      if (isSlotsChildren(child, child.children)) {
        slots = child.children;
        if (child.children.default) {
          _options.push(...travelDropDownChildren(child.children.default()));
        }
      } else if (isArrayChildren(child, child.children)) {
        _options.push(...travelDropDownChildren(child.children));
      }
      options.push({
        _props: omit(props, ["title"]),
        _slots: slots,
        isGroup: true,
        title: props.title,
        options: _options
      });
    } else if (isNamedComponent(child, "Doption")) {
      const props = resolveProps(child);
      let slots;
      if (isSlotsChildren(child, child.children)) {
        slots = child.children;
      }
      options.push({
        _props: props,
        _slots: slots,
        value: (_d = props.value) != null ? _d : getChildrenString((_c = (_b = (_a = child.children).default) == null ? void 0 : _b.call(_a)) != null ? _c : []),
        disabled: props.disabled
      });
    } else if (isNamedComponent(child, "Dsubmenu")) {
      const props = resolveProps(child);
      options.push({
        _props: omit(props, ["trigger", "position", "triggerProps"]),
        isSubmenu: true,
        value: (_h = props.value) != null ? _h : getChildrenString((_g = (_f = (_e = child.children).default) == null ? void 0 : _f.call(_e)) != null ? _g : []),
        disabled: props.disabled,
        footer: child.children.footer,
        render: child.children.default,
        trigger: props.trigger,
        position: props.position,
        children: travelDropDownChildren((_k = (_j = (_i = child.children).content) == null ? void 0 : _j.call(_i)) != null ? _k : [])
      });
    } else if (isArrayChildren(child, child.children)) {
      options.push(...travelDropDownChildren(child.children));
    } else if (isSlotsChildren(child, child.children)) {
      const _children = (_m = (_l = child.children).default) == null ? void 0 : _m.call(_l);
      if (_children) {
        options.push(...travelDropDownChildren(_children));
      }
    } else if (isArray(child)) {
      options.push(...travelDropDownChildren(child));
    }
  }
  return options;
};
export { isGroup, isSubmenu, travelDropDownChildren };
