// @arco-design/web-vue/lib/steps/step.js
"use strict";
var vue = require("vue");
var globalConfig = require("../_utils/global-config.js");
var index = require("../icon/icon-check/index.js");
var index$1 = require("../icon/icon-close/index.js");
var pluginVue_exportHelper = require("../_virtual/plugin-vue_export-helper");
const _sfc_main = vue.defineComponent({
  name: "Step",
  components: {
    IconCheck: index,
    IconClose: index$1
  },
  props: {
    title: String,
    description: String,
    status: {
      type: String
    },
    disabled: {
      type: Boolean,
      default: false
    },
    step: {
      type: Number,
      default: 1
    },
    current: {
      type: Number,
      default: 1
    },
    type: {
      type: String,
      default: "default"
    },
    direction: {
      type: String,
      default: "horizontal"
    },
    labelPlacement: {
      type: String,
      default: "horizontal"
    },
    lineLess: {
      type: Boolean,
      default: false
    },
    nextStepError: {
      type: Boolean,
      default: false
    },
    onClick: {
      type: [Function, Array]
    }
  },
  emits: ["click"],
  setup(props, { emit }) {
    const prefixCls = globalConfig.getPrefixCls("steps-item");
    const iconCls = globalConfig.getPrefixCls("steps-icon");
    const showTail = vue.computed(() => !props.lineLess && (props.labelPlacement === "vertical" || props.direction === "vertical"));
    const handleClick = (ev) => {
      if (!props.disabled) {
        emit("click", props.step, ev);
      }
    };
    const cls = vue.computed(() => [
      prefixCls,
      `${prefixCls}-${props.status}`,
      {
        [`${prefixCls}-active`]: props.step === props.current,
        [`${prefixCls}-next-error`]: props.nextStepError,
        [`${prefixCls}-disabled`]: props.disabled
      }
    ]);
    return {
      prefixCls,
      iconCls,
      cls,
      showTail,
      handleClick
    };
  }
});
function _sfc_render(_ctx, _cache, $props, $setup, $data, $options) {
  const _component_icon_check = vue.resolveComponent("icon-check");
  const _component_icon_close = vue.resolveComponent("icon-close");
  return vue.openBlock(), vue.createElementBlock("div", {
    class: vue.normalizeClass(_ctx.cls),
    onClick: _cache[0] || (_cache[0] = (...args) => _ctx.handleClick && _ctx.handleClick(...args))
  }, [
    _ctx.showTail ? (vue.openBlock(), vue.createElementBlock("div", {
      key: 0,
      class: vue.normalizeClass(`${_ctx.prefixCls}-tail`)
    }, null, 2)) : vue.createCommentVNode("v-if", true),
    _ctx.type !== "arrow" ? (vue.openBlock(), vue.createElementBlock("div", {
      key: 1,
      class: vue.normalizeClass(`${_ctx.prefixCls}-node`)
    }, [
      vue.renderSlot(_ctx.$slots, "node", {
        step: _ctx.step,
        status: _ctx.status
      }, () => [
        _ctx.type !== "dot" ? (vue.openBlock(), vue.createElementBlock("div", {
          key: 0,
          class: vue.normalizeClass(_ctx.iconCls)
        }, [
          vue.renderSlot(_ctx.$slots, "icon", {
            step: _ctx.step,
            status: _ctx.status
          }, () => [
            _ctx.status === "finish" ? (vue.openBlock(), vue.createBlock(_component_icon_check, { key: 0 })) : _ctx.status === "error" ? (vue.openBlock(), vue.createBlock(_component_icon_close, { key: 1 })) : (vue.openBlock(), vue.createElementBlock(vue.Fragment, { key: 2 }, [
              vue.createTextVNode(vue.toDisplayString(_ctx.step), 1)
            ], 2112))
          ])
        ], 2)) : vue.createCommentVNode("v-if", true)
      ])
    ], 2)) : vue.createCommentVNode("v-if", true),
    vue.createElementVNode("div", {
      class: vue.normalizeClass(`${_ctx.prefixCls}-content`)
    }, [
      vue.createElementVNode("div", {
        class: vue.normalizeClass(`${_ctx.prefixCls}-title`)
      }, [
        vue.renderSlot(_ctx.$slots, "default", {}, () => [
          vue.createTextVNode(vue.toDisplayString(_ctx.title), 1)
        ])
      ], 2),
      _ctx.description || _ctx.$slots.description ? (vue.openBlock(), vue.createElementBlock("div", {
        key: 0,
        class: vue.normalizeClass(`${_ctx.prefixCls}-description`)
      }, [
        vue.renderSlot(_ctx.$slots, "description", {}, () => [
          vue.createTextVNode(vue.toDisplayString(_ctx.description), 1)
        ])
      ], 2)) : vue.createCommentVNode("v-if", true)
    ], 2)
  ], 2);
}
var Step = /* @__PURE__ */ pluginVue_exportHelper(_sfc_main, [["render", _sfc_render]]);
module.exports = Step;
