// @arco-design/web-vue/es/upload/upload-progress.js
import { defineComponent, inject, createVNode } from "vue";
import Progress from "../progress/index.js";
import Tooltip from "../tooltip/index.js";
import IconPause from "../icon/icon-pause/index.js";
import IconPlayArrowFill from "../icon/icon-play-arrow-fill/index.js";
import IconUpload from "../icon/icon-upload/index.js";
import IconCheck from "../icon/icon-check/index.js";
import { getProgressStatus } from "./utils.js";
import { getPrefixCls } from "../_utils/global-config.js";
import { useI18n } from "../locale/index.js";
import { uploadInjectionKey } from "./context.js";
var UploadProgress = defineComponent({
  name: "UploadProgress",
  props: {
    file: {
      type: Object,
      required: true
    },
    listType: {
      type: String,
      required: true
    }
  },
  setup(props) {
    const prefixCls = getPrefixCls("upload-progress");
    const {
      t
    } = useI18n();
    const uploadCtx = inject(uploadInjectionKey, void 0);
    const renderIcon = () => {
      var _a, _b, _c, _d;
      if (props.file.status === "error") {
        return createVNode("span", {
          "class": [uploadCtx == null ? void 0 : uploadCtx.iconCls, `${uploadCtx == null ? void 0 : uploadCtx.iconCls}-upload`],
          "onClick": () => uploadCtx == null ? void 0 : uploadCtx.onUpload(props.file)
        }, [(uploadCtx == null ? void 0 : uploadCtx.showRetryButton) && ((_b = (_a = uploadCtx == null ? void 0 : uploadCtx.customIcon) == null ? void 0 : _a.retryIcon) == null ? void 0 : _b.call(_a)) || props.listType === "picture-card" ? createVNode(IconUpload, null, null) : t("upload.retry")]);
      }
      if (props.file.status === "done") {
        return createVNode("span", {
          "class": [uploadCtx == null ? void 0 : uploadCtx.iconCls, `${uploadCtx == null ? void 0 : uploadCtx.iconCls}-success`]
        }, [((_d = (_c = uploadCtx == null ? void 0 : uploadCtx.customIcon) == null ? void 0 : _c.successIcon) == null ? void 0 : _d.call(_c)) || createVNode(IconCheck, null, null)]);
      }
      if (props.file.status === "init") {
        return createVNode(Tooltip, {
          "content": t("upload.start")
        }, {
          default: () => {
            var _a2, _b2;
            return [createVNode("span", {
              "class": [uploadCtx == null ? void 0 : uploadCtx.iconCls, `${uploadCtx == null ? void 0 : uploadCtx.iconCls}-start`],
              "onClick": () => uploadCtx == null ? void 0 : uploadCtx.onUpload(props.file)
            }, [((_b2 = (_a2 = uploadCtx == null ? void 0 : uploadCtx.customIcon) == null ? void 0 : _a2.startIcon) == null ? void 0 : _b2.call(_a2)) || createVNode(IconPlayArrowFill, null, null)])];
          }
        });
      }
      return (uploadCtx == null ? void 0 : uploadCtx.showCancelButton) && createVNode(Tooltip, {
        "content": t("upload.cancel")
      }, {
        default: () => {
          var _a2, _b2;
          return [createVNode("span", {
            "class": [uploadCtx == null ? void 0 : uploadCtx.iconCls, `${uploadCtx == null ? void 0 : uploadCtx.iconCls}-cancel`]
          }, [((_b2 = (_a2 = uploadCtx == null ? void 0 : uploadCtx.customIcon) == null ? void 0 : _a2.cancelIcon) == null ? void 0 : _b2.call(_a2)) || createVNode(IconPause, null, null)])];
        }
      });
    };
    const renderProgress = () => {
      var _a;
      if (["init", "uploading"].includes((_a = props.file.status) != null ? _a : "")) {
        const status = getProgressStatus(props.file.status);
        return createVNode(Progress, {
          "type": "circle",
          "size": "mini",
          "showText": false,
          "status": status,
          "percent": props.file.percent
        }, null);
      }
      return null;
    };
    return () => createVNode("span", {
      "class": prefixCls
    }, [renderProgress(), renderIcon()]);
  }
});
export { UploadProgress as default };
