// @arco-design/web-vue/es/table/table-tr.js
import { defineComponent, computed, openBlock, createElementBlock, normalizeClass, renderSlot } from "vue";
import { getPrefixCls } from "../_utils/global-config.js";
import _export_sfc from "../_virtual/plugin-vue_export-helper";
const _sfc_main = defineComponent({
  name: "Tr",
  props: {
    isExpandRow: {
      type: Boolean
    },
    isEmptyRow: {
      type: Boolean
    },
    checked: {
      type: Boolean
    }
  },
  setup(props) {
    const prefixCls = getPrefixCls("table");
    const cls = computed(() => [
      `${prefixCls}-tr`,
      {
        [`${prefixCls}-tr-expand`]: props.isExpandRow,
        [`${prefixCls}-tr-empty`]: props.isEmptyRow,
        [`${prefixCls}-tr-checked`]: props.checked
      }
    ]);
    return {
      prefixCls,
      cls
    };
  }
});
function _sfc_render(_ctx, _cache, $props, $setup, $data, $options) {
  return openBlock(), createElementBlock("tr", {
    class: normalizeClass(_ctx.cls)
  }, [
    renderSlot(_ctx.$slots, "default")
  ], 2);
}
var Tr = /* @__PURE__ */ _export_sfc(_sfc_main, [["render", _sfc_render]]);
export { Tr as default };
