// @arco-design/web-vue/es/table/table-operation-td.js
import { defineComponent, computed, createVNode } from "vue";
import { getPrefixCls } from "../_utils/global-config.js";
import { getOperationStyle, getOperationFixedCls } from "./utils.js";
import Checkbox from "../checkbox/index.js";
import Radio from "../radio/index.js";
import IconPlus from "../icon/icon-plus/index.js";
import IconMinus from "../icon/icon-minus/index.js";
var OperationTd = defineComponent({
  name: "OperationTd",
  components: {
    Checkbox,
    Radio,
    IconPlus,
    IconMinus
  },
  props: {
    record: {
      type: Object,
      required: true
    },
    rowKey: {
      type: String,
      default: "key"
    },
    operationColumn: {
      type: Object,
      required: true
    },
    operations: {
      type: Array,
      required: true
    },
    isRadio: {
      type: Boolean
    },
    hasExpand: {
      type: Boolean
    },
    selectedRowKeys: {
      type: Array,
      required: true
    },
    expandedIcon: {
      type: Function
    },
    expandedRowKeys: {
      type: Array,
      required: true
    },
    renderExpandBtn: {
      type: Function,
      required: true
    }
  },
  emits: ["select", "expand"],
  setup(props, {
    emit,
    slots
  }) {
    const prefixCls = getPrefixCls("table");
    const style = computed(() => getOperationStyle(props.operationColumn, props.operations));
    const cls = computed(() => [`${prefixCls}-td`, `${prefixCls}-operation`, {
      [`${prefixCls}-checkbox`]: props.operationColumn.name === "selection" && !props.isRadio,
      [`${prefixCls}-radio`]: props.operationColumn.name === "selection" && props.isRadio,
      [`${prefixCls}-expand`]: props.operationColumn.name === "expand"
    }, ...getOperationFixedCls(prefixCls, props.operationColumn)]);
    const renderSelection = () => {
      var _a;
      const rowKey = props.record[props.rowKey];
      if (props.isRadio) {
        return createVNode(Radio, {
          "value": rowKey,
          "modelValue": (_a = props.selectedRowKeys[0]) != null ? _a : "",
          "disabled": Boolean(props.record.disabled),
          "uninjectGroupContext": true,
          "onChange": (value) => emit("select", [value])
        }, null);
      }
      return createVNode(Checkbox, {
        "value": rowKey,
        "modelValue": props.selectedRowKeys,
        "disabled": Boolean(props.record.disabled),
        "uninjectGroupContext": true,
        "onChange": (values) => emit("select", values)
      }, null);
    };
    const renderContent = () => {
      if (props.operationColumn.name === "selection") {
        return renderSelection();
      }
      if (props.operationColumn.name === "expand") {
        if (props.hasExpand) {
          return props.renderExpandBtn(props.record);
        }
        return null;
      }
      if (props.operationColumn.bodyNode) {
        return props.operationColumn.bodyNode(props.record, {
          class: cls.value,
          style: style.value
        });
      }
      return null;
    };
    return () => createVNode("td", {
      "class": cls.value,
      "style": style.value
    }, [createVNode("span", {
      "class": `${prefixCls}-cell`
    }, [renderContent()])]);
  }
});
export { OperationTd as default };
