// @arco-design/web-vue/es/menu/indent.js
import { defineComponent, toRef, openBlock, createElementBlock, Fragment, renderList, normalizeClass, normalizeStyle, createCommentVNode } from "vue";
import useMenuContext from "./hooks/use-menu-context.js";
import { getPrefixCls } from "../_utils/global-config.js";
import _export_sfc from "../_virtual/plugin-vue_export-helper";
const _sfc_main = defineComponent({
  name: "MenuIndent",
  props: {
    level: {
      type: Number,
      default: 1
    }
  },
  setup() {
    const prefixCls = getPrefixCls("menu");
    const menuContext = useMenuContext();
    return {
      prefixCls,
      levelIndent: toRef(menuContext, "levelIndent")
    };
  }
});
const _hoisted_1 = { key: 0 };
function _sfc_render(_ctx, _cache, $props, $setup, $data, $options) {
  return _ctx.level > 1 ? (openBlock(), createElementBlock("span", _hoisted_1, [
    (openBlock(true), createElementBlock(Fragment, null, renderList(_ctx.level - 1, (index) => {
      return openBlock(), createElementBlock("span", {
        key: index,
        class: normalizeClass(`${_ctx.prefixCls}-indent`),
        style: normalizeStyle(`width: ${_ctx.levelIndent}px`)
      }, null, 6);
    }), 128))
  ])) : createCommentVNode("v-if", true);
}
var MenuIndent = /* @__PURE__ */ _export_sfc(_sfc_main, [["render", _sfc_render]]);
export { MenuIndent as default };
