// @arco-design/web-vue/lib/_components/select-view/select-view.js
"use strict";
var vue = require("vue");
var globalConfig = require("../../_utils/global-config.js");
var is = require("../../_utils/is.js");
var inputLabel = require("../input-label/input-label.js");
var index = require("../../input-tag/index.js");
var iconHover = require("../icon-hover.js");
var index$5 = require("../../icon/icon-down/index.js");
var index$2 = require("../../icon/icon-loading/index.js");
var index$1 = require("../../icon/icon-close/index.js");
var index$4 = require("../../icon/icon-expand/index.js");
var index$3 = require("../../icon/icon-search/index.js");
var SelectView = vue.defineComponent({
  name: "SelectView",
  props: {
    modelValue: {
      type: [Object, Array]
    },
    inputValue: String,
    placeholder: String,
    disabled: {
      type: Boolean,
      default: false
    },
    error: {
      type: Boolean,
      default: false
    },
    loading: {
      type: Boolean,
      default: false
    },
    opened: {
      type: Boolean,
      default: false
    },
    size: {
      type: String,
      default: "medium"
    },
    bordered: {
      type: Boolean,
      default: true
    },
    multiple: {
      type: Boolean,
      default: false
    },
    allowClear: {
      type: Boolean,
      default: false
    },
    allowCreate: {
      type: Boolean,
      default: false
    },
    allowSearch: {
      type: Boolean,
      default: (props) => is.isArray(props.modelValue)
    },
    maxTagCount: {
      type: Number,
      default: 0
    },
    formatLabel: {
      type: Function
    },
    retainInputValue: {
      type: Boolean,
      default: false
    }
  },
  emits: ["remove", "clear"],
  setup(props, {
    emit,
    slots,
    attrs
  }) {
    const prefixCls = globalConfig.getPrefixCls("select-view");
    const {
      opened
    } = vue.toRefs(props);
    const componentRef = vue.ref();
    const inputRef = vue.computed(() => {
      var _a;
      return (_a = componentRef.value) == null ? void 0 : _a.inputRef;
    });
    const validValue = vue.computed(() => {
      if (props.multiple && !is.isArray(props.modelValue)) {
        return props.modelValue ? [props.modelValue] : [];
      }
      if (!props.multiple && is.isArray(props.modelValue)) {
        return props.modelValue[0];
      }
      return props.modelValue;
    });
    const isEmptyValue = vue.computed(() => is.isArray(validValue.value) ? validValue.value.length === 0 : !validValue.value);
    const enabledInput = vue.computed(() => props.allowSearch || props.allowCreate);
    const showClearBtn = vue.computed(() => props.allowClear && !props.disabled && !isEmptyValue.value);
    const handleRemove = (tag) => {
      emit("remove", tag);
    };
    const handleClear = (ev) => {
      emit("clear", ev);
    };
    const renderIcon = () => {
      var _a, _b, _c, _d;
      if (props.loading) {
        return (_b = (_a = slots.loadingIcon) == null ? void 0 : _a.call(slots)) != null ? _b : vue.createVNode(index$2, null, null);
      }
      if (props.allowSearch && props.opened) {
        return (_d = (_c = slots.searchIcon) == null ? void 0 : _c.call(slots)) != null ? _d : vue.createVNode(index$3, null, null);
      }
      if (slots.arrowIcon) {
        return slots.arrowIcon();
      }
      if (props.multiple || enabledInput.value) {
        return vue.createVNode(index$4, {
          "style": {
            transform: "rotate(-45deg)"
          }
        }, null);
      }
      return vue.createVNode(index$5, {
        "class": `${prefixCls}-arrow-icon`
      }, null);
    };
    const renderSuffix = () => vue.createVNode(vue.Fragment, null, [showClearBtn.value && vue.createVNode(iconHover, {
      "class": `${prefixCls}-clear-btn`,
      "onClick": handleClear,
      "onMousedown": (e) => e.stopPropagation()
    }, {
      default: () => [vue.createVNode(index$1, null, null)]
    }), vue.createVNode("span", {
      "class": `${prefixCls}-icon`
    }, [renderIcon()])]);
    vue.watch(opened, (opened2) => {
      if (!opened2 && inputRef.value && inputRef.value.isSameNode(document.activeElement)) {
        inputRef.value.blur();
      }
    });
    const cls = vue.computed(() => [`${prefixCls}-${props.multiple ? "multiple" : "single"}`, {
      [`${prefixCls}-opened`]: props.opened,
      [`${prefixCls}-borderless`]: !props.bordered
    }]);
    const render = () => {
      if (is.isArray(validValue.value)) {
        return vue.createVNode(index, {
          "ref": componentRef,
          "baseCls": prefixCls,
          "class": cls.value,
          "modelValue": validValue.value,
          "inputValue": props.inputValue,
          "formatTag": props.formatLabel,
          "focused": props.opened,
          "placeholder": props.placeholder,
          "disabled": props.disabled,
          "size": props.size,
          "error": props.error,
          "maxTagCount": props.maxTagCount,
          "disabledInput": !props.allowSearch && !props.allowCreate,
          "retainInputValue": true,
          "onRemove": handleRemove
        }, {
          prefix: slots.prefix,
          suffix: renderSuffix,
          tag: slots.label
        });
      }
      return vue.createVNode(inputLabel, {
        "ref": componentRef,
        "baseCls": prefixCls,
        "class": cls.value,
        "modelValue": validValue.value,
        "inputValue": props.inputValue,
        "focused": props.opened,
        "placeholder": props.placeholder,
        "disabled": props.disabled,
        "size": props.size,
        "error": props.error,
        "formatLabel": props.formatLabel,
        "enabledInput": enabledInput.value
      }, {
        default: slots.label,
        prefix: slots.prefix,
        suffix: renderSuffix
      });
    };
    return {
      inputRef,
      render
    };
  },
  methods: {
    focus() {
      if (this.inputRef) {
        this.inputRef.focus();
      }
    },
    blur() {
      if (this.inputRef) {
        this.inputRef.blur();
      }
    }
  },
  render() {
    return this.render();
  }
});
module.exports = SelectView;
